### PR TITLE
dockerTools: configurable compression schema

### PR DESCRIPTION
**Pull request by athre0z**
## Description of changes

This PR adds support for swapping out the compression algorithm used in all major dockerTools commands that generate images. The default algorithm remains unchanged (gzip). I also added some new tests to cover the new code paths.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/build-support/docker/default.nix [hunk 1]
```diff
@@ -76,6 +78,30 @@ let
...
+      ext = "";
+      nativeInputs = [ ];
+      compress = "cat";
+      decompress = "cat";
+    };
+    gz = {
+      ext = ".gz";
+      nativeInputs = [ pigz ];
+      compress = "pigz -p$NIX_BUILD_CORES -nTR";
+      decompress = "pigz -d -p$NIX_BUILD_CORES";
```
Comment (athre0z) [inline pkgs/build-support/docker/default.nix:92]: Drive-by fix-up: I also restricted pigz to the `NIX_BUILD_CORES` instead of letting it guess while I moved the command up here. Happy to put this into a separate commit if so desired.

File: pkgs/build-support/docker/default.nix [hunk 2]
```diff
@@ -761,16 +792,27 @@ rec {
   mergeImages = images: runCommand "merge-docker-images"
```
Comment (athre0z) [inline pkgs/build-support/docker/default.nix:792]: This function has fixed arguments and cannot easily be extended to take the new `compressor` argument. To work around this, I simply made it pick the compression schema used in the last image.

File: pkgs/build-support/docker/default.nix [hunk 3]
```diff
@@ -761,16 +792,27 @@ rec {
...
     for item in $images; do
       name=$(basename $item)
       mkdir inputs/$name
-      tar -I pigz -xf $item -C inputs/$name
+
+      last_image_mime=$(file --mime-type -b $item)
+      case $last_image_mime in
+        "application/x-tar") ${compressors.none.decompress};;
+        "application/zstd") ${compressors.zstd.decompress};;
+        "application/gzip") ${compressors.gz.decompress};;
```
Comment (roberth) [inline pkgs/build-support/docker/default.nix:813]:
```suggestion
        "application/gzip") ${compressors.gz.decompress};;
        *) echo "error: unexpected layer type $last_image_mime"; exit 1
          ;;
```

Comment (roberth) [reply pkgs/build-support/docker/default.nix:813]: Needs `-o pipefail` to fail correctly. Do we have that?

File: pkgs/build-support/docker/default.nix [hunk 4]
```diff
@@ -787,7 +829,13 @@ rec {
     mv repositories image/repositories
     mv manifest.json image/manifest.json
     # Create tarball and gzip
-    tar -C image --hard-dereference --sort=name --mtime="@$SOURCE_DATE_EPOCH" --owner=0 --group=0 --xform s:'^./':: -c . | pigz -nTR > $out
+    tar -C image --hard-dereference --sort=name --mtime="@$SOURCE_DATE_EPOCH" --owner=0 --group=0 --xform s:'^./':: -c . | (
+      case $last_image_mime in
+        "application/x-tar") ${compressors.none.compress};;
+        "application/zstd") ${compressors.zstd.compress};;
+        "application/gzip") ${compressors.gz.compress};;
+      esac
```
Comment (roberth) [inline pkgs/build-support/docker/default.nix:837]:
```suggestion
        # `*)` not needed; already checked.
      esac
```

File: pkgs/build-support/docker/default.nix [hunk 5]
```diff
@@ -539,6 +566,8 @@ rec {
       buildVMMemorySize ? 512
     , # Time of creation of the image.
       created ? "1970-01-01T00:00:01Z"
+    , # Compressor to use. One of: none, gz, zstd.
```
Comment (roberth) [inline pkgs/build-support/docker/default.nix:569]: Needs documentation in the manual, `doc/build-helpers/images/dockertools.section.md`

**Review by roberth — CHANGES_REQUESTED** [review]

File: pkgs/build-support/docker/default.nix (hunk 3 above)
Comment (athre0z) [reply pkgs/build-support/docker/default.nix:813]: Yes. I'm not sure where exactly it is specified, but it appears to be enabled by default in `runCommand`. Adding a `set | grep SHELLOPTS` as the first line prints:

```
merge-docker-images> SHELLOPTS=braceexpand:errexit:hashall:interactive-comments:pipefail
```

While running the tests.

**Comment by roberth** [comment]
@ofborg build nixosTests.docker-tools

**Review by roberth — COMMENTED** [review]
Diff lgtm; waiting for ofborg.

**Review by drupol — APPROVED** [review]

**Review by roberth — APPROVED** [review]
